### PR TITLE
CumulusCI `error` Command

### LINKED ISSUE #1494: Proposal: Change error handling
I envision the following future user interface:

    $ cci some command
    Error: Xyzzy
    Type `cci error help` for more information about debugging errors.

    $ cci error help

    If you'd like to dig into this error more yourself, you can get the last few lines 
    of context about it from 'cci error info'.

    If you'd like to submit it to a developer for conversation, you can 
    use the 'cci error gist' command. Just make sure that your GitHub access token 
    has the 'create gist' scope.

    For more info see: https://cumulusci.readthedocs.io/en/latest/features.html#reporting-error-logs
    
    $ cci error info
    You can get the FULL TRACE from <path to full history>. Here are the last <x> lines:

    Traceback (most recent call last):
    File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 218, in main
        cli(standalone_mode=False)
    File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/click/core.py", line 764, in __call__
        return self.main(*args, **kwargs)
    File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/click/core.py", line 717, in main
        rv = self.invoke(ctx)
    File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
        return _process_result(sub_ctx.command.invoke(sub_ctx))
    File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/click/core.py", line 956, in invoke
        return ctx.invoke(self.callback, **ctx.params)
    File "/Users/pprescod/.venvs/CCI-37/lib/python3.7/site-packages/click/core.py", line 555, in invoke
        return callback(*args, **kwargs)
    File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 165, in new_func
        func(RUNTIME, *args, **kw)
    File "/Users/pprescod/code/CumulusCI/cumulusci/cli/cci.py", line 281, in shell
        runpy.run_path(script, init_globals={**globals(), **locals()})
    File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 263, in run_path
        pkg_name=pkg_name, script_name=fname)
    File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 96, in _run_module_code
        mod_name, mod_spec, pkg_name, script_name)
    File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/runpy.py", line 85, in _run_code
        exec(code, run_globals)
    File "foo.py", line 11, in <module>
        foo()
    File "foo.py", line 7, in foo
        1 / 0
    ZeroDivisionError: division by zero
    > /Users/pprescod/code/CumulusCI/foo.py(7)foo()
    -> 1 / 0

====

If the log file has the words "Traceback (most recent call last):" in the last 1000 lines, that's where it should start printing. Otherwise I guess it could just print the last 15 lines.

### LINKED ISSUE #1498: Gist suggestion should appear less often
`cci` should not output a gist suggestion on every error, but under the following circumstances:

1. It has not been output in the last 30 minutes, and
2. The `CCI_DEBUG` environment variable is not set

### PR DESCRIPTION
**Pull request by Br4nd0R**
# Changes
* New top level `cci error` command for interacting with errors in CumulusCI
* `cci gist` moved to `cci error gist`
* `cci error info` displays lines of a stacktrace from the previous `cci` command run. 
* Prompt when encountering error's in cci now suggests `cci error help` (instead of `cci gist`)
* closes #1494 
* closes #1498

### CODE REVIEW

**Comment by Br4nd0R** [comment]
@prescod - Keeping this as a draft as I still need to update documentation and tests, but let me know if this is what you were thinking. 

I also included a num_lines argument to `cci error info` that defaults to 30. Thought it may be nice if the output is clipped off a couple of lines too early that you could display more with `cci error info 50` instead of `cat`'ing your most recent logfile 🤷‍♂

File: cumulusci/cli/cci.py [hunk 1]
```diff
@@ -368,6 +330,23 @@ def service():
...
+    from `cci error info`.
+
+    If you'd like to submit it to a developer for conversation,
+    you can use the `cci error gist` command. Just make sure
+    that your GitHub access token has the 'create gist' scope.
+
+    For more information on working with errors in CumulusCI visit:
+    https://cumulusci.readthedocs.io/en/latest/features.html#working-with-errors
+    """
+    pass
```
Comment (prescod) [inline cumulusci/cli/cci.py]: Python trivia: the docstring does the same thing as pass, so you actually don't need it here (although some might argue that it helps clarify the intent). The ellipsis operator also does a passable job as pass, but I'm sure someone would complain if you used it that way.

Comment (Br4nd0R) [reply cumulusci/cli/cci.py]: TIL!

File: cumulusci/cli/cci.py [hunk 2]
```diff
@@ -1419,3 +1396,75 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
...
+@click.argument("num_lines", required=False)
+def info(num_lines):
+    if not CCI_LOGFILE_PATH.is_file():
+        click.echo(f"No logfile found at: {CCI_LOGFILE_PATH}")
+        return
+
+    if not num_lines:
+        output_lines = 30
+
+    output = lines_from_traceback(CCI_LOGFILE_PATH.read_text(), output_lines)
```
Comment (Br4nd0R) [inline cumulusci/cli/cci.py]: note to self:
```suggestion
    output = lines_from_traceback(CCI_LOGFILE_PATH.read_text(), output_lines or 30)
```

File: cumulusci/cli/cci.py (hunk 1 above)
Comment (jstvz) [reply cumulusci/cli/cci.py]: 👨‍🎓

File: cumulusci/cli/cci.py [hunk 3]
```diff
@@ -1419,3 +1396,75 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
                 "Scratch org deletion failed.  Ignoring the error below to complete the flow:"
             )
             click.echo(str(e))
+
+
+@error.command(name="info", help="Outputs the last part of the most recent traceback")
+@click.argument("num_lines", required=False)
```
Comment (jstvz) [inline cumulusci/cli/cci.py]: I definitely don't think this should be an argument, but rather an option (`-n`/`--num_lines`). Inconsistency in how we're using these across commands means they can be confusing for folks (for example, see `cci org scratch`). 

Also, I'd expect us to expand this command in the future to connect to ErrorsDB.

File: cumulusci/cli/cci.py [hunk 4]
```diff
@@ -1419,3 +1396,75 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
...
+def info(num_lines):
+    if not CCI_LOGFILE_PATH.is_file():
+        click.echo(f"No logfile found at: {CCI_LOGFILE_PATH}")
+        return
+
+    if not num_lines:
+        output_lines = 30
```
Comment (jstvz) [inline cumulusci/cli/cci.py]: If we switch to an option we can define it like:

```python
@click.option('--num_lines', '-n', default=30, show_default=True, type=int)
```

...and won't need to do it here.

File: cumulusci/cli/cci.py [hunk 5]
```diff
@@ -1419,3 +1396,75 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
...
+
+    output = lines_from_traceback(CCI_LOGFILE_PATH.read_text(), output_lines)
+    click.echo(output)
+
+
+def lines_from_traceback(log_content, num_lines):
+    """Returns the the last num_lines of the logfile,
+    or the whole traceback, whichever is shorter. If
+    no stacktrace is found in the logfile, the user is
+    notified.
```
Comment (jstvz) [inline cumulusci/cli/cci.py]: 💭 : I'd think that if I pass `num_lines` I should get the last _n_ lines, in addition to the traceback.

Comment (Br4nd0R) [reply cumulusci/cli/cci.py]: what if the traceback is 300 lines long?

Comment (jstvz) [reply cumulusci/cli/cci.py]: Okay, in that case... maybe `num_lines` is really `max_lines`?

File: cumulusci/cli/cci.py [hunk 6]
```diff
@@ -1419,3 +1395,71 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
...
             click.echo(str(e))
+
+
+@error.command(name="info", help="Outputs the last part of the most recent traceback")
+@click.option("--num-lines", "-n", default=30, show_default=True, type=int)
+def info(num_lines):
+    if not CCI_LOGFILE_PATH.is_file():
+        click.echo(f"No logfile found at: {CCI_LOGFILE_PATH}")
+    else:
+        output = lines_from_traceback(CCI_LOGFILE_PATH.read_text(), num_lines or 30)
```
Comment (jstvz) [inline cumulusci/cli/cci.py]: If `num_lines` isn't passed by the user then it should be equal to the default value, 30.

**Comment by Br4nd0R** [comment]
@jstvz @prescod - Let me know if either of you want additional changes or if this can be approved.

File: cumulusci/cli/cci.py [hunk 7]
```diff
@@ -1419,3 +1395,74 @@ def flow_run(runtime, flow_name, org, delete_org, debug, o, skip, no_prompt):
...
+def error_info(max_lines):
+    if not CCI_LOGFILE_PATH.is_file():
+        click.echo(f"No logfile found at: {CCI_LOGFILE_PATH}")
+    else:
+        output = lines_from_traceback(CCI_LOGFILE_PATH.read_text(), max_lines)
+        click.echo(output)
+
+
+def lines_from_traceback(log_content, num_lines):
+    """Returns the the last num_lines of the logfile,
```
Comment (prescod) [inline cumulusci/cli/cci.py]:
```suggestion
    """Returns the the last max_lines of the logfile,
```

**Comment by prescod** [comment]
I've made one minor suggestion and then I'm ready to approve.

File: docs/features.rst [hunk 8]
```diff
@@ -732,8 +732,20 @@ Caveats:
...
+===================
+Logfiles
+--------
+CumulusCI creates a logfile every time a cci command is run. The only exception to this is when the cci gist command is run. Logfiles are stored under `~/.cumulusci/logs`.
+
+Viewing Stacktraces
+-------------------
+If you encounter an error and want more information on what went wrong, you can use `cci error info` to display the last 30 lines of the stacktrace (if present) from the last command you executed in CumulusCI.
+
+You can include the option num_lines argument if you want to customize how much of the stacktrae you see.
```
Comment (jstvz) [inline docs/features.rst]:
```suggestion
You can include the option num_lines argument if you want to customize how much of the stacktrace you see.
```

File: docs/features.rst [hunk 9]
```diff
@@ -732,8 +732,20 @@ Caveats:
 * See `this link <https://ntotten.com/2018/05/11/convert-metadata-to-source-format-while-maintain-git-history/>`_ for some tips on preserving git history while converting your source format.
 
 
+Working with Errors
+===================
+Logfiles
+--------
+CumulusCI creates a logfile every time a cci command is run. The only exception to this is when the cci gist command is run. Logfiles are stored under `~/.cumulusci/logs`.
```
Comment (jstvz) [inline docs/features.rst]:
```suggestion
CumulusCI creates a log file every time a cci command besides ``gist`` is run. Log files are stored under ``~/.cumulusci/logs``.
```

RST needs double backticks for inline code. 😒

File: docs/features.rst [hunk 10]
```diff
@@ -732,8 +732,20 @@ Caveats:
 * See `this link <https://ntotten.com/2018/05/11/convert-metadata-to-source-format-while-maintain-git-history/>`_ for some tips on preserving git history while converting your source format.
 
 
+Working with Errors
+===================
+Logfiles
```
Comment (jstvz) [inline docs/features.rst]:
```suggestion
Log files
```
I think this is [two words](https://en.wikipedia.org/wiki/Log_file)

File: docs/features.rst [hunk 11]
```diff
@@ -732,8 +732,20 @@ Caveats:
...
+--------
+CumulusCI creates a logfile every time a cci command is run. The only exception to this is when the cci gist command is run. Logfiles are stored under `~/.cumulusci/logs`.
+
+Viewing Stacktraces
+-------------------
+If you encounter an error and want more information on what went wrong, you can use `cci error info` to display the last 30 lines of the stacktrace (if present) from the last command you executed in CumulusCI.
```
Comment (jstvz) [inline docs/features.rst]:
```suggestion
If you encounter an error and want more information on what went wrong, you can use ``cci error info`` to display the last *n* lines (30, by default) of the stacktrace (if present) from the last command you executed in CumulusCI.
```

File: docs/features.rst [hunk 12]
```diff
@@ -732,8 +732,20 @@ Caveats:
...
+===================
+Log Files
+--------
+CumulusCI creates a log file every time a cci command besides ``gist`` is run. Log files are stored under ``~/.cumulusci/logs``.
+
+Viewing Stacktraces
+-------------------
+If you encounter an error and want more information on what went wrong, you can use ``cci error info`` to display the last *n* lines (30, by default) of the stacktrace (if present) from the last command you executed in CumulusCI.
+
+You can include the option num_lines argument if you want to customize how much of the stacktrace you see.
```
Comment (jstvz) [inline docs/features.rst]:
```suggestion
You can include the option ``max_lines`` argument if you want to customize how much of the stacktrace you see.
```

**Comment by davisagli** [comment]
@prescod this is the one I said in standup that I wanted to get acquainted with before merging, and I just talked to @Br4nd0R about some changes I'd like :(

**Comment by davisagli** [comment]
Okay, I've actually looked through this now and think it's mostly good -- but I would prefer showing the full traceback by default.

File: cumulusci/cli/cci.py (hunk 1 above)
Comment (davisagli) [reply cumulusci/cli/cci.py]: Guido's take on the difference between `pass`, `...`, and `None`: https://twitter.com/llanga/status/1216312568131792896

**Comment by Br4nd0R** [comment]
I've gone ahead and made [this issue](https://github.com/SFDO-Tooling/CumulusCI/issues/1514) for the enhancement @davisagli.

**Pull request merged**